flips/context-propagation.md

```diff
@@ -0,0 +1,208 @@
+
+This assumption is no longer true today. The access node, for example, may communicate with multiple (unstaked) consensus followers. It's perceivable that in the future there will be even more cases where communication with unstaked parties may happen (for example, execution nodes talking to DPS).
+
+Furthermore, many engines actually don't need to know the `OriginID` of a message at all. For example, the synchronization engine only needs to know the origin of a sync request so that it can send back a response. However, the more direct form of this information is the peer ID of the libp2p node.
+
+Finally, as of today the network layer does not actually verify the [`OriginID`](https://github.com/onflow/flow-go/blob/2641774451dcc1e147843a74265afb58466b50ab/network/message/message.pb.go#L29) field of messages received from the network. This opens the door to impersonation attacks, as a node can send a message pretending to be any other node simply by putting the other node's `OriginID` field inside the message they send.
+
+## Proposal
+
+### Separate lifecycle management from state observation
+
+Moving forward, we should split the `ReadyDoneAware` interface into two separate interfaces:
+- A `Runnable` interface which can be used to start and stop a component:
+  ```golang
+  type Runnable interface {
+    // Run will start the component
+    Run(context.Context)
+  }
+  ```
+- A `ReadyDoneAware` interface with methods `Ready` and `Done` which can be used to check a component's state (i.e wait for startup / shutdown to complete)
+  ```golang
+  type ReadyDoneAware interface {
+    // Ready returns a channel that will close when component startup has completed.
+    Ready() <-chan struct{}
+    // Done returns a channel that will close when component shutdown has completed.
+    Done() <-chan struct{}
+  }
+  ```
+
+A component will now be started by passing a `Context` to its `Run` method, and can be stopped by cancelling the `Context`. If a component needs to startup subcomponents, it can create child `Context`s from this `Context` and pass those to the subcomponents.
+#### Motivations
+- `Context`s are the standard way of doing go-routine lifecycle management in Go, and adhering to standards helps eliminate confusion and ambiguity for anyone interacting with the `flow-go` codebase. This is especially true now that we are beginning to provide API's and interfaces for third parties to interact with the codebase (e.g DPS).
+  - Even to someone unfamiliar with our codebase (but familiar with Go idioms), it is clear how a method signature like `Run(context.Context)` will behave. A method signature like `Ready()` is not so clear.
+  - If context propagation is done properly, there is no need to worry about any cleanup code in the `Done` method.
```

Comment:
I understand the idea, but `ReadyDoneAware` was plugged because cancelling context doesn't provide a way for component to signalise cleanup. 
Its especially useful where components rely on being shutdown in order - for example. Ledger and WAL and Checkpointers.

Reply:
> I understand the idea, but ReadyDoneAware was plugged because cancelling context doesn't provide a way for component to signalise cleanup.

Yes, and this is precisely why I'm not suggesting we get rid of `ReadyDoneAware`. We should keep it, as a way for components to signal that they have finished startup for finished shutdown / cleanup, but it should *not* also be responsible for actually starting or stopping the component.